### PR TITLE
Implement flake8-async plugin

### PR DESCRIPTION
**Pull request by qdegraaf**
### Summary

This PR adds the flake8-async plugin: https://github.com/cooperlees/flake8-async

First Ruff PR, relatively new to Rust as well so all feedback is welcomed. I checked the contributing guide but if I ended up missing a step anyway let me know.

Closes: https://github.com/charliermarsh/ruff/issues/4246

Marked as draft until:

- [x] Fixed bug where from style imports are not caught

### CODE REVIEW

**Comment by qdegraaf** [comment]
> Looks like #4433 was submitted simultaneously?

Yeah, see Issue discussion at https://github.com/charliermarsh/ruff/issues/4246. 

Started on this about a week ago only mentioning it in Discord while I was still in looking-and-playing-around-a-lot stage of devving and unsure whether I could get it to a proper feature in a reasonable time frame, so made no commitment on the issue. Only finished it up today when https://github.com/charliermarsh/ruff/pull/4433 was also pushed after a mention on the issue, which I then saw after already having opened the PR. 

Happy to close if that implementation is preferred. Unsure what the protocol is in these scenarios.

File: crates/ruff/src/rules/flake8_async/rules.rs
Comment (charliermarsh) [inline crates/ruff/src/rules/flake8_async/rules.rs:201]: @qdegraaf - I changed this to use `resolve_call_path`, which will automatically detect situations like `from os import system` or even `from os import system as foo` or `import os as foo; foo.system`.

File: crates/ruff/src/rules/flake8_async/rules.rs
Comment (charliermarsh) [inline crates/ruff/src/rules/flake8_async/rules.rs:138]: @qdegraaf - I opted to remove the extra struct to house the candidate members, it didn't feel like it was buying us much, and without it we can just do a simple `.contains` check.

File: crates/ruff/src/registry.rs
Comment (charliermarsh) [inline crates/ruff/src/registry.rs:743]: We don't have any codes this long, but I don't see a good reason to deviate from upstream in this case 🤷

**Comment by charliermarsh** [comment]
@qdegraaf - From my perspective this is pretty much good to go -- but let me know if you have any questions or want to handle any follow-ups from your end before I merge.

File: crates/ruff/src/rules/flake8_async/rules.rs
Comment (inikolaev) [inline crates/ruff/src/rules/flake8_async/rules.rs:67]: Adjust these comments since we use `ASYNC` for clarity? 
```suggestion
/// ASYNC100
```

File: crates/ruff/src/rules/flake8_async/rules.rs
Comment (inikolaev) [inline crates/ruff/src/rules/flake8_async/rules.rs:224]: @charliermarsh Do you see this as belonging to some utilities - might reusable in the future? Or how do you usually approach these?

File: crates/ruff/src/rules/flake8_async/rules.rs
Comment (MichaReiser) [inline crates/ruff/src/rules/flake8_async/rules.rs:223]: Does `any` work here?

```suggestion
        .any(|scope| {
            if let ScopeKind::Function(FunctionDef { async_, .. }) = &scope.kind {
                *async_
            } else {
                false
            }
        })
```

File: crates/ruff/src/rules/flake8_async/rules.rs
Comment (MichaReiser) [reply crates/ruff/src/rules/flake8_async/rules.rs:224]: I tend to keep them local to rules until we see to use them in many places. But it could be helpful to have an `in_async` on `context` or `scopes`

**Review by MichaReiser — APPROVED** [review]

**Comment by qdegraaf** [comment]
> @qdegraaf - From my perspective this is pretty much good to go -- but let me know if you have any questions or want to handle any follow-ups from your end before I merge.

Awesome, agree with all the edits, thanks for polishing this up. Feel free to merge at your leisure.

**Pull request merged**
